pkg/lifecycle/resource.go

```diff
@@ -0,0 +1,94 @@
+package lifecycle
+
+import (
+	"sync"
+	"sync/atomic"
+	"unsafe"
+)
+
+// Resource keeps track of references and has some compile time debug hooks
+// to help diagnose leaks. It keeps track of if it is open or not and allows
+// blocking until all references are released.
+type Resource struct {
+	sem  sync.RWMutex
+	open bool
+}
+
+// Open waits for any outstanding references, of which there should be none
+// and marks the reference counter as open.
+func (res *Resource) Open() {
+	res.sem.Lock()
+	res.open = true
+	res.sem.Unlock()
```

Comment:
It's great that these `Unlock()` calls are not deferred. I've been bitten by `defer`s building up over time in a busy process.